x-pack/plugins/apm/server/lib/helpers/get_document_sources.ts

```diff
@@ -17,58 +16,127 @@ export async function getDocumentSources({
+      : []),
+    ApmDocumentType.TransactionMetric as const,
+  ].flatMap((documentType) => {
+    const docTypeConfig = getConfigForDocumentType(documentType);
+
+    return (
+      enableContinuousRollups
+        ? docTypeConfig.rollupIntervals
+        : [RollupInterval.OneMinute]
+    ).flatMap((rollupInterval) => {
+      const searchParams = {
+        apm: {
+          sources: [
+            {
               documentType,
               rollupInterval,
-              hasDocs: response.hits.total.value > 0,
-            };
-          })
-        );
+            },
+          ],
+        },
+        body: {
+          track_total_hits: 1,
+          size: 0,
+          terminate_after: 1,
+        },
+      };
+
+      return {
+        documentType,
+        rollupInterval,
+        before: {
+          ...searchParams,
+          body: {
+            ...searchParams.body,
+            query: {
+              bool: {
+                filter: [...kql, ...beforeRange],
+              },
+            },
+          },
+        },
+        current: {
+          ...searchParams,
+          body: {
+            ...searchParams.body,
+            query: {
+              bool: {
+                filter: [...kql, ...currentRange],
+              },
+            },
+          },
+        },
+      };
+    });
+  });
+
+  const allSearches = sourcesToCheck.flatMap(({ before, current }) => [
+    before,
+    current,
+  ]);
+
+  const allResponses = (
+    await apmEventClient.msearch('get_document_availability', ...allSearches)
+  ).responses;
+
+  const checkedSources = sourcesToCheck.map((source, index) => {
+    const responseBefore = allResponses[index * 2];
+    const responseAfter = allResponses[index * 2 + 1];
+    const { documentType, rollupInterval } = source;
 
-        return allHasDocs;
-      })
-    )
+    const hasDataBefore = responseBefore.hits.total.value > 0;
+    const hasDataAfter = responseAfter.hits.total.value > 0;
+
+    return {
+      documentType,
+      rollupInterval,
+      hasDataBefore,
+      hasDataAfter,
+    };
+  });
+
+  const hasAnyDataBefore = !!checkedSources.find(
+    (source) => source.hasDataBefore
   );
 
+  const sources: Array<ApmDataSource & { hasDocs: boolean }> =
+    checkedSources.map((source) => {
+      const { documentType, hasDataAfter, hasDataBefore, rollupInterval } =
+        source;
+
+      return {
+        documentType,
+        rollupInterval,
+        // If there is any data before, we require that data is available before
+        // this time range to mark this source as available. If we don't do that,
+        // users that upgrade to a version that starts generating service tx metrics
+        // will see a mostly empty screen for a while after upgrading.
+        // If we only check before, users with a new deployment will use raw transaction
+        // events.
+        hasDocs: hasAnyDataBefore
+          ? hasDataBefore
+          : hasDataAfter || hasDataBefore,
```

Comment:
This requires some mental gymnastics. Especially `hasDataAfter || hasDataBefore` is hard to grok.

Reply:
In `: hasDataAfter || hasDataBefore,` isn't `hasDataBefore` always `false`? Well only end up in this branch if `hasAnyDataBefore` meaning there's no data from any sources.

So it can be rewritten to:
```ts
hasDocs: hasAnyDataBefore
          ? hasDataBefore
          : hasDataAfter
```

Reply:
> In : hasDataAfter || hasDataBefore, isn't hasDataBefore always false?

Yes you're right, much better 👍

Reply:
Actually, I'm not sure if that makes it easier to understand, or harder. Because then it requires you to understand that when `hasAnyDataBefore===false`, `hasDataBefore` is always `false` as well. I'll try something else.